### PR TITLE
Filter box match filter props

### PR DESCRIPTION
**Pull request by pochretien**
Review only the code!

Its part of a big change for the MultiSelect

### CODE REVIEW

**Comment by MichelML** [comment]
@pochretien can you describe what the big change is in details so we can better review?

**Comment by pochretien** [comment]
Add the MatchFilter in the State.

The MatchFilter is useful when you want a custom logic for your filter instant of the default behaviour.  You can add as arguments a list of properties in your object to test.

eg :
```
<MultiSelectWithPredicateAndFilter
        id={'multiSelectFilterPredicate'}
        items={this.state.hoc}
        options={defaultFlatSelectOptions}
        matchPredicate={(p: string, i: IItemBoxProps) => this.matchPredicate(p, i)}
/>

private matchPredicate(predicate: string, item: IItemBoxProps) {
        const value = parseInt(item.value, 10);
        if (predicate === defaultFlatSelectOptions[0].id) {
            return true;
        } else if (predicate === defaultFlatSelectOptions[1].id) {
            return value % 2 === 0;
        } else if (predicate === defaultFlatSelectOptions[2].id) {
            return value % 2 === 1;
        } else {
            return true;
        }
    }
```

File: src/components/filterBox/FilterBox.tsx [hunk 1]
```diff
@@ -87,7 +89,7 @@ export class FilterBox extends React.Component<IFilterBoxProps, any> {
 
     componentWillMount() {
         if (this.props.onRender) {
-            this.props.onRender(this.props.id);
+            this.props.onRender(this.props.id, this.props.matchFilter);
         }
```
Comment (gdostie) [inline src/components/filterBox/FilterBox.tsx]: Minor suggestion: 
```ts
callIfDefined(this.props.onRender, this.props.id, this.props.matchFilter);
```

File: src/components/filterBox/FilterBoxActions.ts [hunk 2]
```diff
@@ -6,18 +7,35 @@ export const FilterActions = {
...
 };
 
+export const defaultMatchFilter = (filterValue: string, item: any, propertiesName: string[]) => {
+    if (filterValue === '') {
+        return true;
+    }
+
+    if (filterValue && item) {
+        const regex = new RegExp(filterValue, 'gi');
+        return _.some(propertiesName, (property: string) => item[property] ? regex.test(item[property]) : false);
```
Comment (gdostie) [inline src/components/filterBox/FilterBoxActions.ts]:
```ts
return _.some(propertiesName, (property: string) => item[property] && regex.test(item[property]));
```

File: src/components/filterBox/FilterBoxActions.ts [hunk 3]
```diff
@@ -6,18 +7,35 @@ export const FilterActions = {
...
     id: string;
+    matchFilter?: MatchFilter;
 }
 
 export interface IChangeFilterActionPayload extends IFilterActionPayload {
     filterText: string;
 }
 
-export const addFilter = (id: string): IReduxAction<IFilterActionPayload> => ({
+export const addFilter = (id: string, matchFilter: MatchFilter = defaultMatchFilter): IReduxAction<IFilterActionPayload> => ({
```
Comment (gdostie) [inline src/components/filterBox/FilterBoxActions.ts]: If `defaultMatchFilter`'s type is defined properly, typescript can guess the type by the given default value, so no need to type directly `matchFilter`

File: src/components/filterBox/tests/FilterBox.spec.tsx [hunk 4]
```diff
@@ -47,6 +47,18 @@ describe('FilterBox', () => {
             expect(renderSpy.calls.count()).toBe(1);
         });
 
+        it('should call prop onRender with the id and the matchFilter to set in the store', () => {
+            const renderSpy = jasmine.createSpy('onRender');
+            const matchFilter = jasmine.createSpy('matchFilter');
```
Comment (gdostie) [inline src/components/filterBox/tests/FilterBox.spec.tsx]: `matchFilterSpy`

File: src/components/filterBox/tests/FilterBox.spec.tsx [hunk 5]
```diff
@@ -47,6 +47,18 @@ describe('FilterBox', () => {
...
 
+        it('should call prop onRender with the id and the matchFilter to set in the store', () => {
+            const renderSpy = jasmine.createSpy('onRender');
+            const matchFilter = jasmine.createSpy('matchFilter');
+
+            expect(() => filterBoxInstance.componentWillMount()).not.toThrow();
+
+            filterBox.setProps({id: id, onRender: renderSpy, matchFilter: matchFilter});
+            filterBox.unmount();
+            filterBox.mount();
```
Comment (sroy3) [inline src/components/filterBox/tests/FilterBox.spec.tsx]: You really need to unmount and mount after a setProps? If it doesn't work, maybe try an update after instead

**Pull request closed without merging**